### PR TITLE
kazel: support auto-config shard_count

### LINKED ISSUE #42827: support autoconfig for bazel's shard_count
## Enhancement

support autoconfig for bazel's shard_count.

Because tidb's UT cannot run parallel.  so we need to use bazel's shard_count to isolate and run parallel with different threads. But gazelle cannot generate shard_count. so we need a tools to set it.

### PR DESCRIPTION
**Pull request by hawkingrei**
### What problem does this PR solve?

Issue Number: close #42827

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: tools/tazel/ast.go [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+	path, err := filepath.Abs(path)
+	if err != nil {
+		return err
+	}
+	f, err := parser.ParseFile(fset, path, nil, parser.AllErrors)
+	if err != nil {
+		return err
+	}
+	ast.Inspect(f, func(n ast.Node) bool {
+		// Called recursively.
```
Comment (lance6716) [inline tools/tazel/ast.go:65]: We only need to count top-level TestXXX functions, no need to search for them recursively? Seems looping ast.File.Decls is enough

I often use https://yuroyoro.github.io/goast-viewer/ to get a structured AST

Comment (lance6716) [reply tools/tazel/ast.go:65]: BTW, what's the rule of adjusting shard_count, should we count `TestSuiteXXX` and `(*SuiteXXX) TestYYY` as two functions?

Comment (hawkingrei) [reply tools/tazel/ast.go:65]: I have fixed this problem. Now we only count ```TestSuiteXXX``` and ignore ```(*xxxx)TestYYY``` .

Comment (lance6716) [reply tools/tazel/ast.go:65]: No need to use ast.Inspect, we can only check f.Decls

Comment (hawkingrei) [reply tools/tazel/ast.go]: Done

**Review by lance6716 — APPROVED** [review]

**Review by xhebox — APPROVED** [review]

**Pull request merged**